crypto/storage/azure/keyvault.go

```diff
@@ -0,0 +1,276 @@
+		return nil, "", errors.New("only ES256 keys are supported")
+	}
+	// Code to support all types, supported by Azure Key Vault (kept here for future support):
+	//var signatureAlgorithm azkeys.SignatureAlgorithm
+	//switch *key.Kty {
+	//case azkeys.KeyTypeEC:
+	//	fallthrough
+	//case azkeys.KeyTypeECHSM:
+	//	switch *key.Crv {
+	//	case azkeys.CurveNameP256:
+	//		signatureAlgorithm = azkeys.SignatureAlgorithmES256
+	//	case azkeys.CurveNameP256K:
+	//		signatureAlgorithm = azkeys.SignatureAlgorithmES256K
+	//	case azkeys.CurveNameP384:
+	//		signatureAlgorithm = azkeys.SignatureAlgorithmES384
+	//	case azkeys.CurveNameP521:
+	//		signatureAlgorithm = azkeys.SignatureAlgorithmES512
+	//	default:
+	//		return nil, "", fmt.Errorf("unsupported EC curve: %s", *key.Crv)
+	//	}
+	//case azkeys.KeyTypeRSA:
+	//	fallthrough
+	//case azkeys.KeyTypeRSAHSM:
+	//	signatureAlgorithm = azkeys.SignatureAlgorithmPS256
+	//default:
+	//	return nil, "", fmt.Errorf("unsupported key type: %s", *key.Kty)
+	//}
+	return publicKey, azkeys.SignatureAlgorithmES256, nil
+}
+
+var allowedKeyNameRegex = regexp.MustCompile("[^0-9a-zA-Z-]+")
+
+func keyIDToKeyName(keyID string) string {
+	return allowedKeyNameRegex.ReplaceAllString(keyID, "-")
+}
+
+var _ crypto.Signer = &azureSigningKey{}
+
+type azureSigningKey struct {
+	client           keyVaultClient
+	timeOut          time.Duration
+	keyName          string
+	publicKey        crypto.PublicKey
+	signingAlgorithm azkeys.SignatureAlgorithm
+}
+
+func (a azureSigningKey) Public() crypto.PublicKey {
+	return a.publicKey
+}
+
+func (a azureSigningKey) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), a.timeOut)
+	defer cancel()
+	// Sanity check
+	if opts != nil && opts.HashFunc() == 0 {
+		return nil, errors.New("hashing should've been done")
+	}
+	response, err := a.client.Sign(ctx, a.keyName, "", azkeys.SignParameters{
+		Algorithm: to.Ptr(a.signingAlgorithm),
+		Value:     digest,
+	}, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to sign with Azure Key Vault: %w", err)
+	}
+	// Azure Key Vault returns the signature in a []byte with r and s components concatenated.
+	// We need to convert it to an ASN.1-encoded signature. The first half of the signature is r, the second half is s.
+	// Inspired by ecdsa/ecdsa.go#encodeSignature()
+	builder := &cryptobyte.Builder{}
+	builder.AddASN1(asn1.SEQUENCE, func(child *cryptobyte.Builder) {
+		asn1Int(child, response.Result[:len(response.Result)/2])
+		asn1Int(child, response.Result[len(response.Result)/2:])
+	})
+	return builder.Bytes()
+}
+
+func asn1Int(b *cryptobyte.Builder, bytes []byte) {
+	for bytes[0] == 0 {
+		bytes = bytes[1:]
+	}
+	b.AddASN1(asn1.INTEGER, func(c *cryptobyte.Builder) {
+		if bytes[0]&0x80 != 0 {
+			c.AddUint8(0)
+		}
+		c.AddBytes(bytes)
+	})
+}
```

Comment:
magic

Reply:
Copied it literally and made it clear it's copied. I don't understand the code 100%, but this has to do with signed/unsigned requiring a leading zero. I added this as comment.